### PR TITLE
Split writer from interpreter

### PR DESCRIPTION
**Pull request by PalumboN**
Fix https://github.com/pharo-project/opensmalltalk-vm/issues/356
Fix https://github.com/pharo-project/opensmalltalk-vm/issues/354

I left some comments in the code where I was not sure what is happen.

### CODE REVIEW

File: smalltalksrc/VMMaker/ImageReader.class.st [hunk 1]
```diff
@@ -140,7 +140,7 @@ ImageReader >> readHeaderFrom: f startingAt: headerStart [
 	<var: #header type: #SpurImageHeaderStruct>
 	<returnTypeC: #SpurImageHeaderStruct>
 	| header |
-	self simulationOnly: [ header := SpurImageHeaderStruct new ].
+	self simulationOnly: [ header := SpurImageHeaderStruct new ]. "It's ok this simulationOnly: ?"
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/ImageReader.class.st]: ??

File: smalltalksrc/VMMaker/ImageWriter.class.st [hunk 2]
```diff
@@ -0,0 +1,219 @@
+Class {
+	#name : #ImageWriter,
+	#superclass : #VMClass,
+	#instVars : [
+		'objectMemory',
+		'interpreter'
+	],
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/ImageWriter.class.st:7]: Now _ImageWrite_ needs _Interpreter_ just for success check (`interpreter success: ...`). Maybe we can move that responsibility to a 3rd object.

File: smalltalksrc/VMMaker/ImageWriter.class.st [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	self putLong: header initialSpecialObjectsOop toFile: f.
+	self putLong: header hdrLastHash toFile: f.
+	self putLong: header screenSize toFile: f.
+	self putLong: header headerFlags toFile: f.
+	self putWord32: header extraVMMemory toFile: f.
+	self putShort: header hdrNumStackPages toFile: f.
+	self putShort: header hdrCogCodeSize toFile: f.
+	self putWord32: header hdrEdenBytes toFile: f.
+	self putShort: header hdrMaxExtSemTabSize toFile: f.
+	self putShort: 0 toFile: f. "the2ndUnknownShort"
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/ImageWriter.class.st:154]: This `the2ndUnknownShort` doesn't seem to use. 
This is the implementation from simulated code. What should I do? Delete it or pass it through header?

File: smalltalksrc/VMMaker/ImageWriter.class.st [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+{ #category : #writing }
+ImageWriter >> writeImageFile: header [
+	"Write the image header and heap contents to imageFile for snapshot. c.f. writeImageFileIOSimulation.
+	 The game below is to maintain 64-bit alignment for all putLong:toFile: occurrences."
+	<inline: #never>
+	| headerStart headerSize imageName f imageBytes bytesWritten |
+	<var: #headerStart type: #squeakFileOffsetType>
+	<var: #imageName declareC: 'extern char imageName[]'>
+	<var: #f type: #sqImageFile>
+	<var: #sCWIfn type: #'void *'> "?"
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/ImageWriter.class.st]: Here I have to declare the types for all variables, no? And delete the declarations for already not-defined. I'll need help with that.

File: smalltalksrc/VMMaker/ImageWriter.class.st [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+	headerStart := 0.  
+	headerSize := header headerSize.
+	imageName := interpreter imageName.
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/ImageWriter.class.st]: This `imageName` can come by parameter.

File: smalltalksrc/VMMaker/StackInterpreter.class.st [hunk 6]
```diff
@@ -10553,6 +10561,40 @@ StackInterpreter >> nativeStackSize [
...
+{ #category : #'image save/restore' }
+StackInterpreter >> newHeader [
+
+	<doNotGenerate>
+	| header |
+
+	header := SpurImageHeaderStruct new.
+
+	header imageFormat: self imageFormatVersion.	
+	header swapBytes: false. "TO CHECK, depends of image format"
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/StackInterpreter.class.st]: TO CHECK

Comment (PalumboN) [reply smalltalksrc/VMMaker/StackInterpreter.class.st]: This field is not needed for the writer

File: smalltalksrc/VMMaker/ImageReader.class.st (hunk 1 above)
Comment (tesonep) [reply smalltalksrc/VMMaker/ImageReader.class.st]: Yes, the header is allocated in the stack

Comment (tesonep) [reply smalltalksrc/VMMaker/ImageReader.class.st]:
```suggestion
	self simulationOnly: [ header := SpurImageHeaderStruct new ].
```

File: smalltalksrc/VMMaker/StackInterpreter.class.st [hunk 7]
```diff
@@ -10553,6 +10557,39 @@ StackInterpreter >> nativeStackSize [
 	^ 4 *1024 * 1024
 ]
 
+{ #category : #'image save/restore' }
+StackInterpreter >> newHeader [
+
+	<doNotGenerate>
```
Comment (tesonep) [inline smalltalksrc/VMMaker/StackInterpreter.class.st]:
```suggestion
```

**Pull request merged**